### PR TITLE
implement RoundState.processCertificatesM

### PR DESCRIPTION
**Pull request by haroldcarr**

### CODE REVIEW

**Review by cwjnkins — APPROVED** [review]
If you are OK with the small commit I just pushed, feel free to merge this!

**Comment by haroldcarr** [comment]
merged - thanks

File: LibraBFT/Impl/Consensus/Liveness/RoundState.agda [hunk 1]
```diff
@@ -41,13 +43,28 @@ processLocalTimeoutM now obmEpoch round = do
 
 ------------------------------------------------------------------------------
 
+maybeAdvanceRound : Round → SyncInfo → Maybe (Round × NewRoundReason)
+
 processCertificatesM : Instant → SyncInfo → LBFT (Maybe NewRoundEvent)
-processCertificatesM now syncInfo = do
+processCertificatesM now syncInfo =
+  logEE ("RoundState" ∷ "processCertificatesM" {-∷ lsSI syncInfo-} ∷ []) $ do
   rshcr <- use (lRoundState ∙ rsHighestCommittedRound)
```
Comment (mark-moir) [inline LibraBFT/Impl/Consensus/Liveness/RoundState.agda:51]: This differs from the Haskell code, which directly uses the `rsHighestCommittedRound` lens derived from the `RWRoundState` constraint.  I'm just noting here that I would like to discuss this at higher bandwidth.  I think at a minimum we should have some explanation about this.  I think it makes sense to put the explanation with the definitions of the `l*` lenses (like `lRoundState` in this case).

File: LibraBFT/Impl/Consensus/Liveness/RoundState.agda [hunk 2]
```diff
@@ -41,13 +43,28 @@ processLocalTimeoutM now obmEpoch round = do
...
-  if-RWST (syncInfo ^∙ siHighestCommitRound <? rshcr) -- TODO : define and use 'when'
-    then pure unit -- IMPL-TODO ((lRoundState ∙ rsHighestCommittedRound) :=  (syncInfo ^∙ siHighestCommitRound))
-    else pure unit
-  pure nothing
+  when (syncInfo ^∙ siHighestCommitRound <? rshcr) $ do
```
Comment (mark-moir) [inline LibraBFT/Impl/Consensus/Liveness/RoundState.agda:52]: @haroldcarr : This should be `>?`.

**Review by mark-moir — COMMENTED** [review]
Mostly looks good @haroldcarr, but I have a few comments including one important one that should be addressed.  Sorry I missed the merge train!

File: LibraBFT/Impl/Consensus/Liveness/RoundState.agda [hunk 3]
```diff
@@ -41,13 +43,28 @@ processLocalTimeoutM now obmEpoch round = do
 
 ------------------------------------------------------------------------------
 
+maybeAdvanceRound : Round → SyncInfo → Maybe (Round × NewRoundReason)
+
 processCertificatesM : Instant → SyncInfo → LBFT (Maybe NewRoundEvent)
-processCertificatesM now syncInfo = do
+processCertificatesM now syncInfo =
+  logEE ("RoundState" ∷ "processCertificatesM" {-∷ lsSI syncInfo-} ∷ []) $ do
```
[the remark this answers is not in the document]
Comment (haroldcarr) [reply LibraBFT/Impl/Consensus/Liveness/RoundState.agda:50]: I have commented out all usage of `logEE` - except one in `SafetyRules.constructAndSignVoteM`
To comment that one out, someone needs to adjust contracts (I tried, but it wasn't obvious).

File: LibraBFT/Impl/Consensus/Liveness/RoundState.agda (hunk 2 above)
Comment (haroldcarr) [reply LibraBFT/Impl/Consensus/Liveness/RoundState.agda:52]: Great catch.   I think I know how this happened.

File: LibraBFT/Impl/Consensus/Liveness/RoundState.agda (hunk 1 above)
Comment (haroldcarr) [reply LibraBFT/Impl/Consensus/Liveness/RoundState.agda:51]: I put a comment in LibraBFT.ImplShared.Consensus.Types

**Comment by haroldcarr** [comment]
I addressed this in:

  https://github.com/oracle/bft-consensus-agda/pull/101

**Pull request merged**